engine/util/src/main/scala/pl/touk/nussknacker/engine/util/MathUtils.scala

```diff
@@ -0,0 +1,114 @@
+      })
+    }
+  }
+
+  def min(n1: Number, n2: Number): Number = {
+    implicit val promotionStrategy: NumberTypesPromotionStrategy = NumberTypesPromotionStrategy.ForMinMax
+    withNotNullValues(n1, n2) {
+      withValuesWithTheSameType(n1, n2)(new SameNumericTypeHandler {
+        override def onBytes(n1: Byte, n2: Byte): Byte = Math.min(n1, n2).byteValue()
+        override def onShorts(n1: Short, n2: Short): Short = Math.min(n1, n2).shortValue()
+        override def onInts(n1: Int, n2: Int): Int = Math.min(n1, n2)
+        override def onLongs(n1: Long, n2: Long): Long = Math.min(n1, n2)
+        override def onBigIntegers(n1: BigInteger, n2: BigInteger): BigInteger = n1.min(n2)
+        override def onFloats(n1: Float, n2: Float): Float = Math.min(n1, n2)
+        override def onDoubles(n1: Double, n2: Double): Double = Math.min(n1, n2)
+        override def onBigDecimals(n1: java.math.BigDecimal, n2: java.math.BigDecimal): java.math.BigDecimal = n1.min(n2)
+      })
+    }
+  }
+
+  def sum(n1: Number, n2: Number): Number = {
+    implicit val promotionStrategy: NumberTypesPromotionStrategy = NumberTypesPromotionStrategy.ForMathOperation
+    withNotNullValues(n1, n2) {
+      withValuesWithTheSameType(n1, n2)(new SameNumericTypeHandler {
+        override def onBytes(n1: Byte, n2: Byte): Byte = throw new IllegalStateException("Bytes should be promoted to Ints before addition")
+        override def onShorts(n1: Short, n2: Short): Short = throw new IllegalStateException("Shorts should be promoted to Ints before addition")
+        override def onInts(n1: Int, n2: Int): Int = n1 + n2
+        override def onLongs(n1: Long, n2: Long): Long = n1 + n2
+        override def onBigIntegers(n1: BigInteger, n2: BigInteger): BigInteger = n1.add(n2)
+        override def onFloats(n1: Float, n2: Float): Float = n1 + n2
+        override def onDoubles(n1: Double, n2: Double): Double = n1 + n2
+        override def onBigDecimals(n1: java.math.BigDecimal, n2: java.math.BigDecimal): java.math.BigDecimal = n1.add(n2)
+      })
+    }
+  }
+  
+  private def withNotNullValues(n1: Number, n2: Number)(f: => Number)
+                               (implicit promotionStrategy: NumberTypesPromotionStrategy): Number = {
+    if (n1 == null) {
+      if (n2 == null) null else convertToPromotedType(n2)
+    } else if (n2 == null) {
+      convertToPromotedType(n1)
+    } else {
+      f
+    }
+  }
+
+  private def withValuesWithTheSameType(n1: Number, n2: Number)(handler: SameNumericTypeHandler)
+                                       (implicit promotionStrategy: NumberTypesPromotionStrategy): Number = {
+    val promotedClass = promotionStrategy.promoteClasses(n1.getClass, n2.getClass).asInstanceOf[SingleTypingResult].objType.klass
+    if (promotedClass == classOf[java.lang.Byte]) {
+      handler.onBytes(SpringNumberUtils.convertNumberToTargetClass(n1, classOf[java.lang.Byte]), SpringNumberUtils.convertNumberToTargetClass(n2, classOf[java.lang.Byte]))
```

Comment:
maybe we can write val promoted1 = SpringNumberUtils.convertNumberToTargetClass(n1, promotedClass) just once?
and than do pattern matching?

Reply:
Do you want to pattern mach on (promoted1, promoted2)? In this case we will lost type checking of `convertNumberToTargetClass` that it returns `T` from `Class<T>`. I'm not sure if it will be more mistake friendly... Or maybe you have some other idea?

Reply:
yeah, true, probably there is not ideal solution here...